cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -0,0 +1,482 @@
+   * @brief Destroy the `device_column_view` object.
+   *
+   * @note Does not free the column data, simply free's the device memory
+   * allocated to hold the child views.
+   *---------------------------------------------------------------------------**/
+  void destroy();
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief A mutable, non-owning view of device data as a column of elements
+ * that is trivially copyable and usable CUDA device code.
+ *---------------------------------------------------------------------------**/
+class alignas(16) mutable_column_device_view
+    : public detail::column_device_view_base {
+ public:
+  mutable_column_device_view() = delete;
+  ~mutable_column_device_view() = default;
+  mutable_column_device_view(mutable_column_device_view const&) = default;
+  mutable_column_device_view(mutable_column_device_view&&) = default;
+  mutable_column_device_view& operator=(mutable_column_device_view const&) =
+      default;
+  mutable_column_device_view& operator=(mutable_column_device_view&&) = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Factory to construct a column view that is usable in device memory.
+   *
+   * Allocates and copies views of `soure_view`'s children to device memory to
+   * make them accessible in device code.
+   *
+   * If `source_view.num_children() == 0`, then no device memory is allocated.
+   *
+   * Returns a `std::unique_ptr<mutable_column_device_view>` with a custom
+   *deleter to free the device memory allocated for the children.
+   *
+   * A `mutable_column_device_view` should be passed by value into GPU kernels.
+   *
+   * @param source_view The `column_view` to make usable in device code
+   * @param stream optional, stream on which the memory for children will be
+   * allocated
+   * @return A `unique_ptr` to a `mutable_column_device_view` that makes the
+   *data from `source_view` available in device memory.
+   *---------------------------------------------------------------------------**/
+  static auto create(mutable_column_view source_view, cudaStream_t stream = 0);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns pointer to the base device memory allocation casted to
+   * the specified type.
+   *
+   * @note If `offset() == 0`, then `head<T>() == data<T>()`
+   *
+   * @note It should be rare to need to access the `head<T>()` allocation of
+   *a column, and instead, accessing the elements should be done via
+   *`data<T>()`.
+   *
+   * @tparam The type to cast to
+   * @return T const* Typed pointer to underlying data
+   *---------------------------------------------------------------------------**/
+  template <typename T = void>
+  __host__ __device__ T* head() const noexcept {
+    return const_cast<T*>(detail::column_device_view_base::head());
```

Comment:
Can't the base be mutable and `column_device_view` enforce immutability? Instead of `const_cast`ing things. Considering that we have `mutable_column_view` derived from the column view base, it doesn't even mean the concept of column view has to be immutable.

Reply:
If `column_device_view_base` were mutable, then it would have all of the `const` and non-`const` accessor member functions. Then, when the immutable `column_device_view` inherits from `column_device_view_base`, it would inherit the mutable members, which I would have to remove.

Instead, the base is immutable, and the mutable members can then be layered on top.